test/mocking_utils/filesystem.hpp

```diff
@@ -0,0 +1,241 @@
+// Copyright 2020 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef MOCKING_UTILS__FILESYSTEM_HPP_
+#define MOCKING_UTILS__FILESYSTEM_HPP_
+
+#include <stdio.h>
+#include <string.h>
+#include <sys/stat.h>
+
+#ifndef _WIN32
+#include <sys/types.h>
+#include <dirent.h>
+#else
+#include <windows.h>
+#endif
+
+#include <map>
+#include <string>
+#include <type_traits>
+
+#include "rcutils/macros.h"
+
+#include "patch.hpp"
+
+namespace mocking_utils
+{
+namespace filesystem
+{
+
+/// Platform-independent set of file type constants.
+struct FileTypes
+{
+  static constexpr mode_t REGULAR_FILE = S_IFREG;
+  static constexpr mode_t DIRECTORY = S_IFDIR;
+};
+
+/// Platform-independent set of file permission constants.
+struct Permissions
+{
+#ifndef _WIN32
+  static constexpr mode_t USER_READABLE = S_IRUSR;
+  static constexpr mode_t USER_WRITABLE = S_IWUSR;
+#else
+  static constexpr mode_t USER_READABLE = _S_IREAD;
+  static constexpr mode_t USER_WRITABLE = _S_IWRITE;
+#endif
+};
+
+#if !defined(_WIN32)
+
+/// Helper class for patching the filesystem API.
+/**
+ * \tparam ID Numerical identifier for this patches. Ought to be unique.
+ */
+template<size_t ID>
+class FileSystem
+{
+public:
+  /// Construct mocked filesystem.
+  /**
+   * \param[in] scope Scope target string, using Mimick syntax.
+   *   \see mocking_utils::Patch documentation for further reference.
+   */
+  explicit FileSystem(const std::string & scope)
+  : opendir_mock_(MOCKING_UTILS_PATCH_TARGET(scope, opendir),
+      MOCKING_UTILS_PATCH_PROXY(opendir)),
+#ifndef _GNU_SOURCE
```

Comment:
Also seems like a good candidate for combining with the ifndef block at line 88 below. Line 87 is the only line shared between the two cases.